File: ios/DeepLinkRouter.swift
import Foundation

enum DeepLinkTarget: String {
    case events
    case materials
    case unknown
}

enum DeepLinkStatus: String {
    case success
    case cancel
    case unknown
}

struct DeepLinkRoute {
    let target: DeepLinkTarget
    let status: DeepLinkStatus
    let rawUrl: String
}

enum DeepLinkRouter {
    static func parse(_ url: URL) -> DeepLinkRoute {
        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        let queryItems = components?.queryItems ?? []

        let statusValue = value(for: "status", in: queryItems)?.lowercased()
        let status: DeepLinkStatus
        switch statusValue {
        case "success":
            status = .success
        case "cancel", "canceled":
            status = .cancel
        default:
            status = .unknown
        }

        let path = url.path.lowercased()
        let flow = value(for: "flow", in: queryItems)?.lowercased() ?? ""
        let target: DeepLinkTarget
        if path.contains("event") || flow.contains("event") {
            target = .events
        } else if path.contains("material") || flow.contains("material") || path.contains("checkout") {
            target = .materials
        } else {
            target = .unknown
        }

        return DeepLinkRoute(
            target: target,
            status: status,
            rawUrl: url.absoluteString
        )
    }

    private static func value(for name: String, in queryItems: [URLQueryItem]) -> String? {
        queryItems.first(where: { $0.name.caseInsensitiveCompare(name) == .orderedSame })?.value
    }
}

File: android/app/src/main/java/com/monsoonfire/portal/reference/DeepLinkRouter.kt
package com.monsoonfire.portal.reference

import android.net.Uri

enum class DeepLinkTarget {
    EVENTS,
    MATERIALS,
    KILN,
    PIECES,
    UNKNOWN
}

enum class DeepLinkStatus {
    SUCCESS,
    CANCEL,
    UNKNOWN
}

data class DeepLinkRoute(
    val target: DeepLinkTarget,
    val status: DeepLinkStatus,
    val rawUrl: String
)

object DeepLinkRouter {
    fun parse(uri: Uri): DeepLinkRoute {
        val statusValue = uri.getQueryParameter("status")?.lowercase().orEmpty()
        val status = when (statusValue) {
            "success" -> DeepLinkStatus.SUCCESS
            "cancel", "canceled" -> DeepLinkStatus.CANCEL
            else -> DeepLinkStatus.UNKNOWN
        }

        val path = uri.path?.lowercase().orEmpty()
        val target = when {
            path.contains("/events") -> DeepLinkTarget.EVENTS
            path.contains("/materials") || path.contains("/store") -> DeepLinkTarget.MATERIALS
            path.contains("/kiln") -> DeepLinkTarget.KILN
            path.contains("/pieces") -> DeepLinkTarget.PIECES
            else -> DeepLinkTarget.UNKNOWN
        }

        return DeepLinkRoute(
            target = target,
            status = status,
            rawUrl = uri.toString()
        )
    }
}
